File: golang/binary_tree_right_side_view.go
func rightSideView(root *TreeNode) []int {
	if root == nil {
		return []int{}
	}

	levels := map[int][]int{}

	helper(root, 1, levels)

	// right side view is just last element of each level
	view := []int{}
	for level := 1; level <= len(levels); level++ {
		vals := levels[level]
		view = append(view, vals[len(vals)-1])
	}

	return view
}

func helper(node *TreeNode, level int, levels map[int][]int) {
	// dfs but we keep track of level and add to levels map
	if node == nil {
		return
	}

	levels[level] = append(levels[level], node.Val)
	helper(node.Left, level+1, levels)
	helper(node.Right, level+1, levels)
}

File: typescript/binary_tree_right_side_view.ts
function rightSideView(root: TreeNode | null): number[] {
	let view = []
	helper(root, 0, view)
	return view
}

function helper(node: TreeNode, level: number, view: number[]) {
	// base case
	if (node == null) return

	// first occurrence right level
	if (level >= view.length) {
		view.push(node.val)
	}

	// dfs right side first
	helper(node.right, level + 1, view)
	helper(node.left, level + 1, view)
}
